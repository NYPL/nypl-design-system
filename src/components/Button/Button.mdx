import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";
import Link from "../Link/Link";

<Meta of={ButtonStories} />

# Button

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.0.4`      |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#button-groups" target="_self">Button Groups</Link>}
- {<Link href="#with-icons" target="_self">With Icons</Link>}
- {<Link href="#patterns" target="_self">Patterns</Link>}
- {<Link href="#other-examples" target="_self">Other Examples</Link>}
- {<Link href="#sizes" target="_self">Sizes</Link>}

## Overview

<Description of={ButtonStories} />

Update the `buttonType` in the Controls section to see the types of buttons
that can be rendered.

- `"primary"` is used for actions that move the user forward. This is the default.
- `"secondary"` is used for actions that move the user back, such as cancellations.
- `"text"` is used for equally weighted actions in a text based list.
- `"callout"` is used for call to action text such as "Donate".
- `"noBrand"` is a variant used when there is no brand and will display the
  background color as black.
- `"pill"` is used for equally weighted actions within cards and always in a set.
- `"link"` has been deprecated and should not be used.

When one and only one `Icon` component is passed inside a `Button` component with
no text, it will automatically be configured to use the `"iconOnly"` type.

## Component Props

<Canvas of={ButtonStories.WithControls} />

<ArgTypes of={ButtonStories.WithControls} />

## Accessibility

The Reservoir `Button` component renders an HTML `<button>` element that is
accessible. `Button`s should be used to perform user actions or submit forms,
never to link and go to another page. For `Button`s without text and only an
icon, make sure to pass an `aria-label` value to the `Button`. `Button`s cannot
have images of text.

`Button`s can be rendered in three sizes through the `size` prop: `"small"`,
`"medium"`, and `"large"`. The default size is `"medium"`. Even though the `"small"`
size is available, we recommend to use the `"medium"` and `"large"` sizes.

The `Button` component also has a `screenreaderOnlyText` prop that can be used to
add additional text to the component that is only visible to screen readers. This
can be used to provide additional context to the `Button` when the text is short,
such as visible "Read more..." text.

Resources:

- [W3C WAI ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button)
- [W3C WAI ARIA Button Examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html)
- [MDN ARIA: button role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role)
- [Chakra UI Button](https://chakra-ui.com/docs/components/form/button)
- [W3C WAI WCAG Target Size](https://www.w3.org/WAI/WCAG21/Understanding/target-size.html)

## Button Groups

Primary and secondary buttons should be grouped like in the following example.
The `ButtonGroup` component should be used to wrap the `Button` component.

<Canvas of={ButtonStories.ButtonGroups} />

## With Icons

The `Icon` component collaborates great with the `Button` component. Check out
[`src/components/Icons/Icon.tsx`](https://github.com/NYPL/nypl-design-system/blob/development/src/components/Icons/Icon.tsx)
to see the full list of icon names that can be rendered.

The icon can be placed to the left or the right of the button text.

<Canvas of={ButtonStories.TextWithIcon} />

Text in a button is optional if an icon is rendered, but make sure that there
is an `aria-label` passed to the `Button` component `aria-label="Previous"`.

<Canvas of={ButtonStories.IconOnly} />

## Patterns

Buttons are blue by default but should be red when it is intended to be used as
a call out. The "Back to Top" button must be of `buttonType="secondary"` and
must include an up arrow icon on the right side.

<Canvas of={ButtonStories.Patterns} />

## Other Examples

The variations modified by the `buttonType` prop:

<Canvas of={ButtonStories.ButtonTypes} />

Modifying the `isDisabled` prop:

<Canvas of={ButtonStories.DisabledButtons} />

## Sizes

Update the size of the `Button` component with the `size` prop and values:
`"small"`, `"medium"`, `"large"`. The default size is `"medium"`.

<Canvas of={ButtonStories.ButtonSizes} />
