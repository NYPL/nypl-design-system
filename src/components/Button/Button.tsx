import * as React from "react";
import bem from "../../utils/bem";
import { ButtonTypes } from "./ButtonTypes";

interface ButtonProps {
  /** Additional attributes passed to the button */
  attributes?: { [key: string]: any };
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** The kind of button assigned through the `ButtonTypes` enum  */
  buttonType?: ButtonTypes;
  /** Additional className for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  className?: string;
  /** Adds 'disabled' property to the button */
  disabled?: boolean;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Trigger the Button's action through the `mouseDown` event handler instead of `onClick`. `false` by default. */
  mouseDown?: boolean;
  /** The action to perform on the `<button>`'s onClick function */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /** The html button attribute */
  type?: "submit" | "button" | "reset";
}

/** Renders a simple `button` element with custom classes and modifiers. */
export default class Button extends React.Component<ButtonProps, any> {
  static defaultProps = {
    mouseDown: false,
  };

  constructor(props: ButtonProps) {
    super(props);
  }

  render(): JSX.Element {
    const {
      attributes,
      blockName,
      buttonType,
      children,
      className,
      disabled,
      id,
      modifiers = [],
      mouseDown,
      onClick,
      type = "submit",
    } = this.props;

    const buttonModifiers = modifiers ? modifiers : [];
    if (!buttonModifiers.find(modifier => modifier === buttonType)) {
      buttonModifiers.push(buttonType);
    }

    const baseClass = "button";

    const btnCallback = mouseDown
      ? { onMouseDown: onClick }
      : { onClick: onClick };

    return (
      <button
        id={id}
        className={`${baseClass} ${bem(baseClass, buttonModifiers, blockName, [
          className,
        ])}`}
        type={type}
        disabled={disabled}
        {...attributes}
        {...btnCallback}
      >
        {children}
      </button>
    );
  }
}
