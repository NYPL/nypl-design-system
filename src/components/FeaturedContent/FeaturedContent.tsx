import { Box, chakra, useMultiStyleConfig } from "@chakra-ui/react";
import React, { forwardRef } from "react";
import { ImageProps } from "../Image/Image";

export const featuredContentWidthArray = [
  "oneQuarter",
  "oneThird",
  "oneHalf",
  "twoThirds",
  "threeQuarters",
];
export type featuredContentWidthType = typeof featuredContentWidthArray[number];

export const featuredContentPositionArray = ["start", "end"] as const;
export type featuredContentPositionType =
  typeof featuredContentPositionArray[number];

interface FeaturedContentImageProps extends ImageProps {
  /** String value that specifies the width of the image rendered within the component. */
  width?: featuredContentWidthType;
  /** String value that specifies the position of the image rendered within the component. */
  position?: featuredContentPositionType;
}

export interface FeaturedContentProps {
  /** Optional CSS class name to add. */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes. */
  id?: string;
  /** The text content rendered in the component.  DS components and native HTML can be passed in this prop. */
  textContent: string | JSX.Element;
  /** Whether component will fill the full width of the browser window, instead of just its parent element.
   * False by default. */
  isFullWidth: boolean;
  /** Data object that contains the props related to the image element: alt, position, src, width.  */
  imageProps: FeaturedContentImageProps;
}

/**
 * The FeaturedContentImage component is used internally in the `FeaturedContent` component. It
 * renders an `Image` component but with overriding styles specific to the
 * `FeaturedContent` component.
 */
function FeaturedContentImage(
  props: React.ComponentProps<"img"> & FeaturedContentImageProps
) {
  const { alt, src } = props;
  return <img alt={alt} src={src} style={{ display: "none" }} />;
}

export const FeaturedContent = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<FeaturedContentProps>>(
    (props) => {
      const {
        textContent,
        isFullWidth,
        imageProps = {
          alt: "",
          position: "end",
          width: "oneHalf",
          src: "",
        },
      } = props;
      const styles = useMultiStyleConfig("FeaturedContent", {
        imagePosition: imageProps.position,
        imageWidth: imageProps.width,
        isFullWidth,
      });

      if (!imageProps) {
        console.warn(
          `NYPL Reservoir FeaturedContent: The 'imageProps' prop is required.`
        );
      } else {
        if (!imageProps.src) {
          console.warn(
            `NYPL Reservoir FeaturedContent: A value for 'imageProps.src' is required.`
          );
        }
        if (!imageProps.alt) {
          console.warn(
            `NYPL Reservoir FeaturedContent: A value for 'imageProps.alt' is required.`
          );
        }
      }

      if (!textContent) {
        console.warn(
          `NYPL Reservoir FeaturedContent: The 'textContent' prop is required.`
        );
      }

      return (
        <Box data-testid="featuredcontent" __css={styles.wrapper}>
          <Box
            data-testid="bg-image"
            __css={{
              ...styles.imgWrapper,
              backgroundImage: `url(${imageProps.src})`,
            }}
          >
            <FeaturedContentImage
              alt={imageProps.alt}
              src={imageProps.src ? imageProps.src : undefined}
            />
          </Box>
          <Box __css={styles.text}>{textContent}</Box>
        </Box>
      );
    }
  )
);

export default FeaturedContent;
