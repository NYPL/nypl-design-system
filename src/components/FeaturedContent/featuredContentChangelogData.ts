/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "Prerelease",
    version: "Prerelease",
    type: "Update",
    affects: ["Styles"],
    notes: [
      'Updated the spacing in the "fullScreen" variant to better align the component text content with the page text content.',
    ],
  },
  {
    date: "2023-10-18",
    version: "2.1.0",
    type: "New Feature",
    affects: ["Accessibility", "Documentation", "Functionality", "Styles"],
    notes: ["Adds FeaturedContent component."],
  },
];
