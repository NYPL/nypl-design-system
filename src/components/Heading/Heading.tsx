// MT-82, MT 225, etc
import * as React from "react";
import bem from "../../utils/bem";
import { DisplaySizes } from "./HeadingDisplaySizes";

export interface HeadingProps {
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** Optional className that appears in addition to `heading` */
  className?: string;
  /** Optional size used to override the default styles of the semantic HTML `<h>` elements */
  displaySize?: DisplaySizes;
  /** Optional ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Number 1-6; used to create the `<h*>` tag */
  level: number;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Inner text of the `<h*>` element */
  text?: string;
  /** Optional URL that header points to; when `url` prop is passed to `Heading`, a child `<a>` element is created and the heading text becomes an active link */
  url?: string;
  /** Optional className for the URL when the `url` prop is passed */
  urlClass?: string;
}

export default function Heading(props: React.PropsWithChildren<HeadingProps>) {
  const {
    blockName,
    className,
    displaySize,
    id,
    level,
    modifiers = [],
    text,
    url,
    urlClass,
  } = props;

  const baseClass = "heading";

  if (displaySize) {
    modifiers.push(displaySize);
  }

  if (level < 1 || level > 6) {
    throw new Error("Heading only supports levels 1-6");
  }

  if (!props.children && !text) {
    throw new Error("Heading has no children, please pass prop: text");
  }

  if (React.Children.count(props.children) > 1) {
    const children = React.Children.map(
      props.children,
      child => (child as JSX.Element).type
    );
    // Catching the error because React's error isn't as helpful.
    throw new Error(
      `Please only pass one child into Heading, got ${children.join(", ")}`
    );
  }

  let content: string | React.ReactNode;
  if (props.children) {
    content = url
      ? React.createElement(
          "a",
          { href: url, className: urlClass },
          props.children
        )
      : props.children;
  } else {
    content = url
      ? React.createElement("a", { href: url, className: urlClass }, text)
      : text;
  }

  const headingProps = {
    className: bem(baseClass, modifiers, blockName, [className]),
    id: id,
  };

  return React.createElement("h" + level, headingProps, content);
}
