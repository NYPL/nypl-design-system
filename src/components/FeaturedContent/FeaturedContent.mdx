import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";
import { changelogData } from "./featuredContentChangelogData";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import * as FeaturedContentStories from "./FeaturedContent.stories";
import Link from "../Link/Link";

<Meta of={FeaturedContentStories} />

# Featured Content

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `2.1.0`    |
| Latest            | `3.2.0`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#component-width-variations" target="_self">Component Width Variations</Link>}
- {<Link href="#image-position-variations" target="_self">Image Position Variations</Link>}
- {<Link href="#image-width-variations" target="_self">Image Width Variations</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

The `FeaturedContent` component provides a method to visually emphasize a text
block with an image, within a full page layout.

## Component Props

<Canvas of={FeaturedContentStories.WithControls} />

<ArgTypes of={FeaturedContentStories.WithControls} />

## Accessibility

The `FeaturedContent` component combines a text block and an image. The text
block (`textContent`) can be any JSX element, so accessibility for child input
elements should follow the accessibility requirements specified for each input
component (for example,
[Button](https://nypl.github.io/nypl-design-system/reservoir/v1/?path=/docs/components-form-elements-button--docs)).
All images must have an `alt` attribute, even if it's empty. The `alt` prop
should be used to concisely describe the image. If the image is decorative, then
the `alt` prop should be an empty string `""`.

Resources for text accessibility:

- [Chakra UI Text](https://chakra-ui.com/docs/components/typography/text)

Resources for image accessibility:

- [W3C WAI Images Tutorial](https://www.w3.org/WAI/tutorials/images/)
- [WebAIM Accessible Images](https://webaim.org/techniques/images/)
- [Yale Usability & Web Accessibility](https://usability.yale.edu/web-accessibility/articles/images)

## Component Width Variations

The width variations of the `FeaturedContent` component can be rendered through
the `isFullWidth` prop. The default is `isFullWidth = false`, and the component
will fill only its parent. If `isFullWidth = true`, component will fill the
screen's width (max 1280px), breaking out of its parent container. This full
layout is best viewed on a full screen, not as shown below.

<Canvas of={FeaturedContentStories.LayoutVariations} />

## Image Position Variations

The image position variations of the `FeaturedContent` component can be rendered
through the `imageProps.position` prop. The default is `end`, and the image will
appear after the text block in a row (on mobile, it will appear below). The
other option is `start`, where the image will appear before the text block in
the row on desktop (and above it on mobile).

<Canvas of={FeaturedContentStories.ImagePositionVariations} />

## Image Width Variations

The image width variations of the `FeaturedContent` component can be rendered
through the `imageProps.width` prop. The options for width are `oneQuarter`,
`oneThird`, `oneHalf`, `twoThirds`, and `threeQuarters`. The default width is
`oneHalf`.

<Canvas of={FeaturedContentStories.imageWidthVariations} />

## Text Content

`textContent` is a string or JSX element passed into `FeaturedContent`, so
accessibility standards and spacing should be considered independently of the
`FeaturedContent` styling.

The above example with an overline, a heading, a short paragraph of body text,
and a CTA button is the recommended usage. While other configurations are
possible, keep in mind that the minimum height of the component is 320px and it
is best to avoid excessive white space.

<Canvas of={FeaturedContentStories.textContentVariations} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
