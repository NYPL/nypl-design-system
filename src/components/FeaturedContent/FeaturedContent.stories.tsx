import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";
import Button from "../Button/Button";
import Heading from "../Heading/Heading";
import Text from "../Text/Text";

import FeaturedContent from "./FeaturedContent";

const meta: Meta<typeof FeaturedContent> = {
  title: "Components/Basic Elements/FeaturedContent",
  component: FeaturedContent,
  decorators: [withDesign],
  argTypes: {
    className: { control: false },
    id: { control: false },
    fullLayout: {
      table: { defaultValue: { summary: false } },
    },
    imageProps: { disable: false },
    "imageProps.alt": {
      control: { type: "text" },
      table: { defaultValue: { summary: "" } },
    },
    "imageProps.width": {
      control: { type: "text" },
      table: { defaultValue: { summary: "oneHalf" } },
    },
    "imageProps.position": {
      control: { type: "text" },
      table: { defaultValue: { summary: "end" } },
    },
    "imageProps.src": {
      table: { defaultValue: { summary: "" } },
    },
  },
};

export default meta;
type Story = StoryObj<typeof FeaturedContent>;

/**
 * Main Story for the FeaturedContent component. This must contain the `args`
 * and `parameters` properties in this object.
 */

export const WithControls: Story = {
  args: {
    className: undefined,
    id: "FeaturedContent-id",
    fullLayout: false,
    //textContent: (
    // <div>
    //   <Button id="test" />
    //   <p>
    //     Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
    //     eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
    //     minim veniam, quis nostrud exercitation ullamco laboris nisi ut
    //     aliquip ex ea commodo consequat. Duis aute irure dolor in
    //     reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
    //     pariatur. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed
    //     do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
    //     ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
    //     aliquip ex ea commodo consequat. Duis aute irure dolor in
    //     reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
    //     pariatur. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed
    //     do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
    //     ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
    //     aliquip ex ea commodo consequat. Duis aute irure dolor in
    //     reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
    //     pariatur.
    //   </p>
    // </div>
    //),
    "imageProps.alt": "Alt text",
    "imageProps.position": "end",
    "imageProps.width": "default",
    "imageProps.src": "//placekitten.com/600/600",
  },
  render: (args) => (
    <FeaturedContent
      className={args.className}
      id={args.id}
      fullLayout={args.fullLayout}
      textContent={
        <div>
          <br />
          <Text size="overline1"> Featured </Text>
          <Heading level="h1"> Spotlight on the Public Domain </Heading>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam Lorem ipsum dolor sit amet, consectetur adipiscing
            elit, sed do eiusmod tempor incididunt ut labore et dolore magna
            aliqua. Ut enim ad minim veniam
          </p>
          <Button id="test"> Discover more </Button>
        </div>
      }
      imageProps={{
        alt: args["imageProps.alt"],
        width: args["imageProps.width"],
        position: args["imageProps.position"],
        src: args["imageProps.src"],
      }}
    ></FeaturedContent>
  ),
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?type=design&node-id=80613-80958&mode=design&t=HerPVZJXMF8fVMUb-0",
    },
    jest: ["FeaturedContent.test.tsx"],
  },
};
