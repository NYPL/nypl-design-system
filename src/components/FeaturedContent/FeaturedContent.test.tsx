import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import FeaturedContent from "./FeaturedContent";
import Button from "../Button/Button";
import Heading from "../Heading/Heading";
export const textContent = (
  <div>
    <br />
    <Heading level="h1" overline="Featured">
      Spotlight on the Public Domain
    </Heading>
    <p>
      The New York Public Library recently enhanced access to all public domain
      items in Digital Collections so that everyone has the freedom to enjoy and
      reuse these materials in almost limitless ways.
    </p>
    <Button id="test"> Discover more </Button>
  </div>
);
const defaultImageProps = {
  alt: "Image example",
  src: "//loremflickr.com/800/400/new+york+public+library",
};

describe("FeaturedContent accessibility tests", () => {
  it("passes for default layout", async () => {
    const { container } = render(
      <FeaturedContent
        isFullWidth={false}
        imageProps={defaultImageProps}
        textContent={textContent}
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
  it("passes for full layout", async () => {
    const { container } = render(
      <FeaturedContent
        isFullWidth={true}
        imageProps={defaultImageProps}
        textContent={textContent}
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("FeaturedContent", () => {
  it("renders the given image and text block", () => {
    render(
      <FeaturedContent
        isFullWidth={true}
        imageProps={defaultImageProps}
        textContent={textContent}
      />
    );

    expect(
      screen.getByText("Spotlight on the Public Domain")
    ).toBeInTheDocument();
    expect(screen.getByTestId("featuredcontent-bg-image")).toBeInTheDocument();
    expect(screen.getByTestId("featuredcontent-bg-image")).toHaveStyle(
      "background-image: url(//loremflickr.com/800/400/new+york+public+library);"
    );
  });

  it("logs a warning if `textContent` prop is not present", () => {
    const warn = jest.spyOn(console, "warn");
    render(
      // @ts-ignore: Typescript complains when a required prop is not passed, but
      // here we don't want to pass the required prop to make sure the warning appears.
      <FeaturedContent isFullWidth={false} imageProps={defaultImageProps} />
    );
    expect(warn).toHaveBeenCalledWith(
      `NYPL Reservoir FeaturedContent: The 'textContent' prop is required.`
    );
  });

  it("logs a warning if `imageProps.src` is empty", () => {
    const warn = jest.spyOn(console, "warn");
    render(
      <FeaturedContent
        isFullWidth={true}
        imageProps={{
          alt: "Image example",
          src: "",
        }}
        textContent={textContent}
      />
    );
    expect(warn).toHaveBeenCalledWith(
      `NYPL Reservoir FeaturedContent: A value for 'imageProps.src' is required.`
    );
  });

  it("logs a warning if `imageProps.alt` is empty", () => {
    const warn = jest.spyOn(console, "warn");
    render(
      <FeaturedContent
        isFullWidth={true}
        imageProps={{
          alt: "",
          src: "//loremflickr.com/800/400/new+york+public+library",
        }}
        textContent={textContent}
      />
    );
    expect(warn).toHaveBeenCalledWith(
      `NYPL Reservoir FeaturedContent: A value for 'imageProps.alt' is required.`
    );
  });
});

it("Renders the UI snapshot correctly", () => {
  const defaultLayout = renderer
    .create(
      <FeaturedContent
        isFullWidth={false}
        imageProps={defaultImageProps}
        textContent={textContent}
      />
    )
    .toJSON();
  const fullLayout = renderer
    .create(
      <FeaturedContent
        isFullWidth={true}
        imageProps={defaultImageProps}
        textContent={textContent}
      />
    )
    .toJSON();

  const oneQuarterImageWidth = renderer
    .create(
      <FeaturedContent
        isFullWidth={false}
        imageProps={{
          alt: "Image example",
          src: "//loremflickr.com/800/400/new+york+public+library",
          width: "oneQuarter",
        }}
        textContent={textContent}
      />
    )
    .toJSON();
  const oneThirdImageWidth = renderer
    .create(
      <FeaturedContent
        isFullWidth={false}
        imageProps={{
          alt: "Image example",
          src: "//loremflickr.com/800/400/new+york+public+library",
          width: "oneThird",
        }}
        textContent={textContent}
      />
    )
    .toJSON();

  const twoThirdsImageWidth = renderer
    .create(
      <FeaturedContent
        isFullWidth={false}
        imageProps={{
          alt: "Image example",
          src: "//loremflickr.com/800/400/new+york+public+library",
          width: "twoThirds",
        }}
        textContent={textContent}
      />
    )
    .toJSON();
  const threeQuartersImageWidth = renderer
    .create(
      <FeaturedContent
        isFullWidth={false}
        imageProps={{
          alt: "Image example",
          src: "//loremflickr.com/800/400/new+york+public+library",
          width: "threeQuarters",
        }}
        textContent={textContent}
      />
    )
    .toJSON();
  expect(defaultLayout).toMatchSnapshot();
  expect(fullLayout).toMatchSnapshot();
  expect(oneQuarterImageWidth).toMatchSnapshot();
  expect(oneThirdImageWidth).toMatchSnapshot();
  expect(twoThirdsImageWidth).toMatchSnapshot();
  expect(threeQuartersImageWidth).toMatchSnapshot();
});
