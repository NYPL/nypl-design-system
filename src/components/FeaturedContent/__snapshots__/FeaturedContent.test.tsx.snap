// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the UI snapshot correctly 1`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the UI snapshot correctly 2`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the UI snapshot correctly 3`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the UI snapshot correctly 4`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the UI snapshot correctly 5`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the UI snapshot correctly 6`] = `
<div
  className="css-0"
  data-testid="featuredcontent"
>
  <div
    className="css-0"
  >
    <div
      className="css-1gt4hjv"
      data-testid="featuredcontent-bg-image"
    >
      <div
        className="css-0"
      >
        <img
          alt="Image example"
          className="css-0"
          src="//placekitten.com/800/400"
        />
      </div>
    </div>
    <div
      className="css-0"
    >
      <div>
        <br />
        <hgroup
          aria-roledescription="Heading group"
          className="css-0"
          role="group"
        >
          <p
            aria-roledescription="Overline"
            className="chakra-text css-rwceap"
            role="paragraph"
          >
            Featured
          </p>
          <h1
            className="chakra-heading css-eyyo6e"
          >
            Spotlight on the Public Domain
          </h1>
        </hgroup>
        <p>
          The New York Public Library recently enhanced access to all public domain items in Digital Collections so that everyone has the freedom to enjoy and reuse these materials in almost limitless ways.
        </p>
        <button
          className="chakra-button css-1xdhyk6"
          id="test"
          type="button"
        >
           Discover more 
        </button>
      </div>
    </div>
  </div>
</div>
`;
